src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java

```diff
@@ -16,9 +25,291 @@
+         * @return this builder
+         * @throws NullPointerException if {@code origin} is {@code null}
+         */
+        public RootTransformBuilder origin(Vector3f origin)
+        {
+            this.originVec = Preconditions.checkNotNull(origin, "Origin must not be null");
+            this.origin = null;
+            return this;
+        }
+
+        /**
+         * Sets the origin of the root transform.
+         *
+         * @param origin the origin name
+         * @return this builder
+         * @throws NullPointerException if {@code origin} is {@code null}
+         * @throws IllegalArgumentException if {@code origin} is not {@code center}, {@code corner} or {@code opposing-corner}
+         */
+        public RootTransformBuilder origin(TransformOrigin origin)
+        {
+            this.origin = Preconditions.checkNotNull(origin, "Origin must not be null");;
+            this.originVec = null;
+            return this;
+        }
+
+        /**
+         * Finish configuring the parent builder
+         * @return the parent block model builder
+         */
+        public BlockModelBuilder end()
+        {
+            return BlockModelBuilder.this;
+        }
+
+        public JsonObject toJson() {
+            // Write the transform to an object
+            JsonObject transform = new JsonObject();
+
+            if (!translation.equals(Vector3f.ZERO))
+            {
+                transform.add("translation", writeVec3(translation));
+            }
+
+            if (!scale.equals(ONE))
+            {
+                transform.add("scale", writeVec3(scale));
+            }
+
+            if (!leftRotation.equals(Quaternion.ONE))
+            {
+                transform.add("rotation", writeQuaternion(leftRotation));
+            }
+
+            if (!rightRotation.equals(Quaternion.ONE))
+            {
+                transform.add("post_rotation", writeQuaternion(rightRotation));
+            }
+
+            if (origin != null)
+            {
+                transform.addProperty("origin", origin.getSerializedName());
+            }
+            else if (originVec != null && !originVec.equals(Vector3f.ZERO))
+            {
+                transform.add("origin", writeVec3(originVec));
+            }
+
+            return transform;
+        }
+
+        public enum TransformOrigin implements StringRepresentable
+        {
+            CENTER(new Vector3f(), "origin"),
+            CORNER(new Vector3f(1f, 1f, 1f), "corner"),
+            OPPOSING_CORNER(new Vector3f(.5f, .5f, .5f), "opposing_corner");
+
+            private final Vector3f vec;
+            private final String name;
+
+            TransformOrigin(Vector3f vec, String name)
+            {
+                this.vec = vec;
+                this.name = name;
+            }
+
+            public Vector3f getVector()
+            {
+                return vec;
+            }
+
+            @Override
+            public String getSerializedName()
```

Comment:
Mark this method as `@NotNull` so the IDEA warning is silenced.
```suggestion
            @Override
            @NotNull
            public String getSerializedName()
```